File: javascript/src/billing/postgres/repositories/preferences.ts
import { z } from "zod";
import type { QueryFn } from "../../../shared/postgres-types.js";
import { StoreError } from "../../../errors.js";
import type { BillingPreferences } from "../../types/index.js";
import {
  optionalRecordRow,
  pgBoolean,
  postgresUuid,
  requireResultField,
  safeParse,
} from "../../../shared/postgres-validation.js";

const BillingPreferencesRowSchema = z
  .object({
    subject_id: postgresUuid,
    auto_recharge: pgBoolean,
    overage_protection: pgBoolean,
    email_notifications: pgBoolean,
    usage_alerts: pgBoolean,
    invoice_reminders: pgBoolean,
  })
  .strict();

export class BillingPreferencesRepository {
  constructor(private query: QueryFn) {}

  async get(userId: string): Promise<BillingPreferences | null> {
    const rows = await this.query("SELECT * FROM bursar.get_billing_preferences($1::uuid)", [
      userId,
    ]);
    const row = optionalRecordRow(rows, "BillingPreferencesRepository.get");
    if (row === null) {
      return null;
    }
    const parsed = safeParse(
      BillingPreferencesRowSchema,
      {
        subject_id: row.subject_id,
        auto_recharge: row.auto_recharge,
        overage_protection: row.overage_protection,
        email_notifications: row.email_notifications,
        usage_alerts: row.usage_alerts,
        invoice_reminders: row.invoice_reminders,
      },
      "BillingPreferencesRepository.get",
    );
    return {
      userId: parsed.subject_id,
      autoRecharge: parsed.auto_recharge,
      overageProtection: parsed.overage_protection,
      emailNotifications: parsed.email_notifications,
      usageAlerts: parsed.usage_alerts,
      invoiceReminders: parsed.invoice_reminders,
    };
  }

  async upsert(prefs: BillingPreferences): Promise<void> {
    const rows = await this.query(
      "SELECT bursar.upsert_billing_preferences($1::uuid, $2, $3, $4, $5, $6) AS updated",
      [
        prefs.userId,
        prefs.autoRecharge,
        prefs.overageProtection,
        prefs.emailNotifications,
        prefs.usageAlerts,
        prefs.invoiceReminders,
      ],
    );
    const updated = requireResultField(
      rows,
      "updated",
      pgBoolean,
      "BillingPreferencesRepository.upsert",
    );
    if (!updated) {
      throw new StoreError("BillingPreferencesRepository.upsert: update was rejected", {
        indeterminate: true,
      });
    }
  }
}

File: python/src/bursar/billing/postgres/repositories/preferences.py
from __future__ import annotations

from uuid import UUID

from bursar.billing.types import BillingPreferences
from bursar.credits.postgres.repositories._types import DbQuery
from bursar.credits.postgres.repositories._utils import (
    optional_mapping_row,
    require_boolean_result,
    validate_non_empty,
)
from bursar.errors import StoreError


class BillingPreferencesRepository:
    def __init__(self, execute: DbQuery) -> None:
        self._execute = execute

    def get(self, user_id: str) -> BillingPreferences | None:
        validate_non_empty(user_id, "user_id")
        rows = self._execute("SELECT * FROM bursar.get_billing_preferences(%s::uuid)", [user_id])
        row = optional_mapping_row(rows, "BillingPreferencesRepository.get")
        if row is None or row.get("subject_id") is None:
            return None
        try:
            subject_id = str(UUID(str(row["subject_id"])))
        except (AttributeError, TypeError, ValueError) as error:
            raise StoreError("BillingPreferencesRepository.get: malformed subject identifier") from error
        boolean_fields = (
            "auto_recharge",
            "overage_protection",
            "email_notifications",
            "usage_alerts",
            "invoice_reminders",
        )
        if any(type(row.get(field)) is not bool for field in boolean_fields):
            raise StoreError("BillingPreferencesRepository.get: malformed boolean field")
        return BillingPreferences(
            user_id=subject_id,
            auto_recharge=row["auto_recharge"],
            overage_protection=row["overage_protection"],
            email_notifications=row["email_notifications"],
            usage_alerts=row["usage_alerts"],
            invoice_reminders=row["invoice_reminders"],
        )

    def upsert(self, prefs: BillingPreferences) -> None:
        user_id = prefs.user_id
        validate_non_empty(user_id, "user_id")
        rows = self._execute(
            "SELECT bursar.upsert_billing_preferences(%s::uuid, %s, %s, %s, %s, %s) AS updated",
            [
                user_id,
                prefs.auto_recharge,
                prefs.overage_protection,
                prefs.email_notifications,
                prefs.usage_alerts,
                prefs.invoice_reminders,
            ],
        )
        if not require_boolean_result(rows, "updated", "BillingPreferencesRepository.upsert"):
            raise StoreError(
                "BillingPreferencesRepository.upsert: update was rejected",
                indeterminate=True,
            )
